apiserver/server_billable_handler.go

```diff
@@ -35,118 +42,66 @@ func BillableEventsHandler(store eventio.BillableEventReader, consolidatedStore
 		storeCtx, cancel := context.WithCancel(context.Background())
 		defer cancel()
 
-		// query the store
-		rowOfRows := RowOfRows{}
-		defer rowOfRows.Close()
-
 		months, err := filter.SplitByMonth()
 		if err != nil {
 			return err
 		}
+
+		delim := ""
 		for _, monthFilter := range months {
-			isConsolidated, err := consolidatedStore.IsRangeConsolidated(monthFilter)
+			err := func () error {  // so we can use defer in-loop
+				isConsolidated, err := consolidatedStore.IsRangeConsolidated(monthFilter)
+				if err != nil {
+					return err
+				}
+				var rows eventio.BillableEventRows
+				if isConsolidated {
+					rows, err = consolidatedStore.GetConsolidatedBillableEventRows(storeCtx, monthFilter)
+				} else {
+					rows, err = store.GetBillableEventRows(storeCtx, monthFilter)
+				}
+				if err != nil {
+					return err
+				}
+				defer rows.Close()
+
+				next := rows.Next()
+				for next {
+					b, err := rows.EventJSON()
+					if err != nil {
+						return err
+					}
+					if !sentOKHeader {
+						// this is sent as late as possible because any errors encountered after
+						// this won't be communicated to the client correctly
+						if err := sendOKHeader(); err != nil {
+							return err
+						}
+					}
+					if _, err := c.Response().Write([]byte(delim)); err != nil {
+						return err
+					}
+					if _, err := c.Response().Write(b); err != nil {
+						return err
+					}
+					delim = ",\n"
+					next = rows.Next()
+					c.Response().Flush()
+				}
+				return nil
+			}()
 			if err != nil {
 				return err
 			}
-			var rows eventio.BillableEventRows
-			if isConsolidated {
-				rows, err = consolidatedStore.GetConsolidatedBillableEventRows(storeCtx, monthFilter)
-			} else {
-				rows, err = store.GetBillableEventRows(storeCtx, monthFilter)
-			}
-			if err != nil {
+		}
+		if !sentOKHeader {
+			if err := sendOKHeader(); err != nil {
 				return err
 			}
-			rowOfRows.RowsCollection = append(rowOfRows.RowsCollection, rows)
 		}
-
-		// stream response to client
-		c.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSONCharsetUTF8)
-		c.Response().WriteHeader(http.StatusOK)
-		return WriteRowsAsJson(c.Response(), c.Response(), &rowOfRows)
-	}
-}
-
-func WriteRowsAsJson(writer io.Writer, flusher http.Flusher, rows eventio.BillableEventRows) error {
-	if _, err := writer.Write([]byte("[\n")); err != nil {
-		return err
-	}
-	flusher.Flush()
-	next := rows.Next()
-	delim := "\n"
-	for next {
-		b, err := rows.EventJSON()
-		if err != nil {
+		if _, err := c.Response().Write([]byte("\n]\n")); err != nil {
```

Comment:
What's this magic byte we're writing?

Reply:
It's the closing `]` of the json array - the top level array doesn't use `encoding/json` because that isn't compatible with streaming responses.

Streaming responses are a bit dubious in general IMHO.